File: jsimple-oauth/src/main/java/jsimple/oauth/extractors/BaseStringExtractorImpl.java
package jsimple.oauth.extractors;

import jsimple.oauth.exceptions.OAuthParametersMissingException;
import jsimple.oauth.model.OAuthRequest;
import jsimple.oauth.model.ParameterList;
import jsimple.oauth.utils.OAuthEncoder;

/**
 * Default implementation of {@link BaseStringExtractor}. Conforms to OAuth 1.0a
 *
 * @author Pablo Fernandez
 */
public class BaseStringExtractorImpl implements BaseStringExtractor {
    /**
     * {@inheritDoc}
     */
    public String extract(OAuthRequest request) {
        checkPreconditions(request);
        String verb = OAuthEncoder.encode(request.getVerb());
        String url = OAuthEncoder.encode(request.getSanitizedUrl());
        String params = getSortedAndEncodedParams(request);
        return verb + "&" + url + "&" + params;
    }

    private String getSortedAndEncodedParams(OAuthRequest request) {
        ParameterList params = new ParameterList();
        params.addAll(request.getQueryStringParams());
        params.addAll(request.getBodyParams());
        params.addAll(new ParameterList(request.getOauthParameters()));
        return params.sort().asOauthBaseString();
    }

    private void checkPreconditions(OAuthRequest request) {
        if (request.getOauthParameters() == null || request.getOauthParameters().size() <= 0)
            throw new OAuthParametersMissingException(request);
    }
}

File: jsimple-oauth/c#/jsimple/oauth/extractors/BaseStringExtractorImpl.cs
namespace jsimple.oauth.extractors {

    using OAuthParametersMissingException = jsimple.oauth.exceptions.OAuthParametersMissingException;
    using OAuthRequest = jsimple.oauth.model.OAuthRequest;
    using ParameterList = jsimple.oauth.model.ParameterList;
    using OAuthEncoder = jsimple.oauth.utils.OAuthEncoder;

    /// <summary>
    /// Default implementation of <seealso cref="BaseStringExtractor"/>. Conforms to OAuth 1.0a
    /// 
    /// @author Pablo Fernandez
    /// </summary>
    public class BaseStringExtractorImpl : BaseStringExtractor {
        /// <summary>
        /// {@inheritDoc}
        /// </summary>
        public virtual string extract(OAuthRequest request) {
            checkPreconditions(request);
            string verb = OAuthEncoder.encode(request.Verb);
            string url = OAuthEncoder.encode(request.SanitizedUrl);
            string @params = getSortedAndEncodedParams(request);
            return verb + "&" + url + "&" + @params;
        }

        private string getSortedAndEncodedParams(OAuthRequest request) {
            ParameterList @params = new ParameterList();
            @params.addAll(request.QueryStringParams);
            @params.addAll(request.BodyParams);
            @params.addAll(new ParameterList(request.OauthParameters));
            return @params.sort().asOauthBaseString();
        }

        private void checkPreconditions(OAuthRequest request) {
            if (request.OauthParameters == null || request.OauthParameters.Count <= 0)
                throw new OAuthParametersMissingException(request);
        }
    }

}